File: src/core/net/encoding/decode-body.ts
import { REPLACEMENT_CHAR, decodeGb18030 } from './gb18030'

const GBK_CHARSETS = new Set(['gbk', 'gb18030', 'gb2312', 'x-gbk', 'csgb2312', 'gb_2312-80'])
const UTF8_CHARSETS = new Set(['utf-8', 'utf8', 'unicode-1-1-utf-8'])

/** 从 `Content-Type` 里抠 charset，小写返回；没有则 null。 */
export function parseCharset(contentType: string | null | undefined): string | null {
  if (!contentType) return null
  const match = /charset\s*=\s*"?([\w-]+)"?/i.exec(contentType)
  return match ? match[1]!.toLowerCase() : null
}

function decodeUtf8(bytes: Uint8Array): string {
  // Hermes 只支持 utf-8 这一种 TextDecoder，正好够用
  return new TextDecoder('utf-8').decode(bytes)
}

function stripBom(text: string): string {
  return text.charCodeAt(0) === 0xfeff ? text.slice(1) : text
}

function countReplacements(text: string): number {
  let count = 0
  for (let i = 0; i < text.length; i++) {
    if (text.charCodeAt(i) === 0xfffd) count += 1
  }
  return count
}

/**
 * 按 API 文档 §0.5 解码响应体：优先信 `Content-Type` 声明的 charset，
 * 未声明时回落 GB18030（MNGA 的做法）。
 *
 * 与 MNGA 不同的是未声明时先试 UTF-8：本项目跟 MNGA 一样带 `__inchst=UTF8`，
 * 服务端多数时候确实返回 UTF-8 却不声明 charset，硬按 GB18030 解会整篇乱码。
 * 判据是替换字符谁少用谁——GBK 中文按 UTF-8 解几乎必然出现大量 U+FFFD，反之亦然。
 */
export function decodeResponseBody(
  bytes: Uint8Array,
  contentType?: string | null,
): string {
  const charset = parseCharset(contentType)
  if (charset && GBK_CHARSETS.has(charset)) return stripBom(decodeGb18030(bytes))
  if (charset && UTF8_CHARSETS.has(charset)) return stripBom(decodeUtf8(bytes))

  const utf8 = decodeUtf8(bytes)
  if (!utf8.includes(REPLACEMENT_CHAR)) return stripBom(utf8)

  const gbk = decodeGb18030(bytes)
  return stripBom(countReplacements(gbk) < countReplacements(utf8) ? gbk : utf8)
}

File: app/src/main/kotlin/com/chasel/ng2n/core/net/encoding/DecodeBody.kt
package com.chasel.ng2n.core.net.encoding

private val GBK_CHARSETS = setOf("gbk", "gb18030", "gb2312", "x-gbk", "csgb2312", "gb_2312-80")
private val UTF8_CHARSETS = setOf("utf-8", "utf8", "unicode-1-1-utf-8")

private val CHARSET_PATTERN = Regex("charset\\s*=\\s*\"?([\\w-]+)\"?", RegexOption.IGNORE_CASE)

/** 从 `Content-Type` 里抠 charset,小写返回;没有则 null。 */
fun parseCharset(contentType: String?): String? {
  if (contentType.isNullOrEmpty()) return null
  val match = CHARSET_PATTERN.find(contentType) ?: return null
  return match.groupValues[1].lowercase()
}

/**
 * WHATWG 的 `TextDecoder('utf-8')` 会**自己吃掉一个前导 BOM**(`ignoreBOM` 默认 false),
 * 外层 [stripBom] 再吃一个 —— 所以 RN 版的 UTF-8 路径最多剥两个 BOM,GBK 路径只剥一个。
 * JVM 的 `String(bytes, UTF_8)` 不剥,这里补上,免得「两个 BOM」这种脏样本两边不一致。
 */
private fun decodeUtf8(bytes: ByteArray): String = stripBom(String(bytes, Charsets.UTF_8))

private fun stripBom(text: String): String = if (text.isNotEmpty() && text[0] == '\uFEFF') text.substring(1) else text

private fun countReplacements(text: String): Int = text.count { it == REPLACEMENT_CHAR }

/**
 * 按 API 文档 §0.5 解码响应体:优先信 `Content-Type` 声明的 charset,
 * 未声明时回落 GB18030(MNGA 的做法)。
 *
 * 与 MNGA 不同的是未声明时先试 UTF-8:本项目跟 MNGA 一样带 `__inchst=UTF8`,
 * 服务端多数时候确实返回 UTF-8 却不声明 charset,硬按 GB18030 解会整篇乱码。
 * 判据是替换字符谁少用谁——GBK 中文按 UTF-8 解几乎必然出现大量 U+FFFD,反之亦然。
 *
 * 四条分支一条都不能少(移植风险 TOP5 #3,漏一处 = 「偶发乱码」级难查 bug):
 * 1. 声明了 GBK 家族 → GB18030;
 * 2. 声明了 UTF-8 家族 → UTF-8;
 * 3. **不认识的 charset 不硬用**,退回投票;
 * 4. 未声明 → 先 UTF-8,一个 U+FFFD 都没有就直接采纳;有才解一遍 GB18030 投票,
 *    替换字符少的胜出,**平手留 UTF-8**。
 * 最后一律剥 BOM。
 */
fun decodeResponseBody(bytes: ByteArray, contentType: String? = null): String {
  val charset = parseCharset(contentType)
  if (charset != null && charset in GBK_CHARSETS) return stripBom(decodeGb18030(bytes))
  if (charset != null && charset in UTF8_CHARSETS) return stripBom(decodeUtf8(bytes))

  val utf8 = decodeUtf8(bytes)
  if (!utf8.contains(REPLACEMENT_CHAR)) return stripBom(utf8)

  val gbk = decodeGb18030(bytes)
  return stripBom(if (countReplacements(gbk) < countReplacements(utf8)) gbk else utf8)
}
